python/ray/data/dataset.py

```diff
@@ -257,6 +262,10 @@ def map_batches(
         batch_size: Optional[int] = 4096,
         compute: Union[str, ComputeStrategy] = None,
         batch_format: str = "native",
+        fn_args: Optional[List[Any]] = None,
```

Comment:
This is half of the parameter list, how does it compare to wrapping them in a FnParams class? Not sure if it's a good idea in Python, but it's good practice in C++ (e.g. https://abseil.io/tips/173).

Reply:
Yeah threading these individual arg collections is annoying, unfortunately there are a few reasons that keeping these as independent arguments is required:

1. When passing these to remote functions (actor or plain tasks), these need to be top-level arguments in order for object refs to be properly resolved to their concrete values, which requires destructuring them before passing them into the functions.
2. When passing these to user UDFs, we need to destructure them in order to cleanly support the default case of a UDF that only takes a `batch` argument: by destructuring these passthrough arguments when calling the UDF, if not are specified, it will be equivalent to a `udf(batch)` call; if you passed them directly as e.g. `fn_args` or a `FnParams` object, if the UDF isn't expecting that argument, then this will fail.
3. When passing these to the block transformation function in the compute strategy, we need to have it result in no arguments passed to block transformation functions that don't support arg passthroughs (e.g. all APIs other than `.map_batches()`). This wouldn't work if we were passing in an explicit `FnParams` object.

Basically this boils down to wanting Datasets operations and UDFs being able to opt-in to receiving these passthrough args, and the passthrough args needing to be top-level arguments in the (actor) tasks.

Reply:
Sg, thanks for the detailed explanation!